### DISCUSSION TITLE
Euler angles and rotation matrix in ComputeElasticityTensor

### DISCUSSION THREAD

**Discussion by Kavan-Khaledi** [idaholab/moose#23957] [Q&A Modules: Solid mechanics] [answered]
Dear Moose team,

I use "ComputeElasticityTensor" to simulate the elastic deformation around a circular tunnel in a transversely isotropic rock (layered rock). The geometry of the model is shown schematically below.
![image](https://user-images.githubusercontent.com/110385277/229809631-90a45be6-ef50-458d-b1d8-3b0fd5d6c1bc.png)

 In this example, the isotropic plane is tilted 45 degrees about the y-axis and the rock medium has an initial isotropic stress of 30 MPa. With these assumptions, we can define the orientation of roke layers by assuming "euler_angle_1=90" and "euler_angle_2=45" (the input file is attached).
My first thought was that the "euler_angle_3" doesn't play a role here, since it only describes the rotation of the isotropic plane (around the axis of anisotropy). But I saw that changing the euler_angle_3 has a significant effect on the deformation pattern around the tunnel. Figure below shows the displacement magnitude for three different values of euler_angle_3:

![image](https://user-images.githubusercontent.com/110385277/229809082-d6a4dd92-841e-4b38-bced-63c2ed83da5f.png)

Only the results for euler_angle_3=90 look reasonable! However, a small change in the ComputeElasticityTensor.C file produces exactly the results I expect.

I changed:
    RotationTensor R(_Euler_angles); // Type R: RealTensorValue.
     // Rotate the elasticity tensor
     _Cijkl.rotate(R);

to:
    RotateTensor R(_Euler_angles); // type R: RealTensorValue.
     // Rotate elasticity tensor
     _Cijkl.rotate(R.transpose());

With the above change in the rotation matrix, the elastic deformation becomes independent of euler_angle_3 for transversely isotropic materials. 

![image](https://user-images.githubusercontent.com/110385277/229811662-9f4cc76a-46ec-48c7-b7c2-5d03848df57a.png)

 
Could you please review the rotation matrix in the ComputeElasticityTensor.C file and let me know if such a change is needed or not? i also attached the input file (+mesh)

Best regards,
Kavan 
[TransverselyIsotropic.zip](https://github.com/idaholab/moose/files/11149248/TransverselyIsotropic.zip)

**Comment by GiudGiud** [comment]
@tonkmr @jessecarterMOOSE on the elasticity tensor

**Comment by jessecarterMOOSE** [comment, marked as answer]
Gotta be really careful here to get what you want. This has tripped me up many times. MOOSE uses a specific Euler angle convention ([there are a lot](https://en.wikipedia.org/wiki/Euler_angles)) which assumes both specific axes being rotated about and the ordering. Be sure to read the documentation on this ([link](https://mooseframework.inl.gov/source/materials/ComputeElasticityTensor.html)). It matters if you're rotating the _material_ to a specific orientation or if you're rotating the _axes_. I believe the default in MOOSE is to use the _latter_. Your modification is likely assuming the former, which is why you want the transpose. There is an example in the documentation similar to this where the _negative_ of the desired angle is needed, which amounts to a transposed rotation matrix (for a single rotation).

If I'm understanding your problem, you want to rotate about the y-axis, but MOOSE uses the Z-X-Z convention for Euler angles 1, 2, and 3. It may be easier in your case to instead provide the _rotation matrix_ (or the transpose of it) to `ComputeElasticityTensor` (again see the docs on how to do this). Or you could re-orient your problem so the desired rotation is about the z-axis.

**Comment by Kavan-Khaledi** [comment]
Dear @jessecarterMOOSE,
Thank you very much for your answer and useful hints. As far as I know, MOOSE uses the Z-X'-Z'' convention (this is also clear from RotationTensor.C). Actually, I wanted to adopt this approach to define the orientation of geological structures (faults, stratifications, rock layers ...) with respect to the global coordinate. In engineering geology, any plane rotated in the x-y-z coordinate is described by only two rotation angles: 1) the dip direction (rotation about Z) and 2) the dip angle (rotation about X'). In other words, the third rotation about Z'' does not change the principal material coordinate. If I understand you correctly, the Euler angle approach may not be directly applicable to these applications. I guess I will follow your suggestion and provide a rotation matrix resulting from only two rotations (Z and X').
Thank you again, Kavan

**Reply by jessecarterMOOSE** [reply]
The ordering of the rotations may still be an issue here (see intrinsic vs extrinsic rotations on the Wikipedia page for Euler angles). Maybe what you want is this?
```
euler_angle_1 = 0
euler_angle_2 = 45
euler_angle_3 = 90
```

**Reply by Kavan-Khaledi** [reply]
Exactly! I changed the order of the rotations from 1-2-3 to 3-2-1. Now the material response (deformation pattern around the tunnel) has become independent of the last rotation (euler_angle_1). This is exactly what I expected for a transversely isotropic material ;) thank you very much for your help.

**Reply by HoWilgh** [reply]
Dear @jessecarterMOOSE ,
In the documentation it is only stated that extrinsic rotation has to be used for euler-angles input. Does the matrix also have to be extrinsic?
Thanks in advance,
HoWil

**Reply by jessecarterMOOSE** [reply]
I believe the rotation matrix, $R$ that you enter gets applied to the elasticity tensor, $C$, directly like: $RCR^T$, so it is up to you to generate the proper rotation matrix

**Reply by HoWilgh** [reply]
To my knowledge a rotation matrix does also hold the information if the rotation is extrinsic or intrinsic. So the results for the rotated tensor and e.g. the resulting strain will be different if you insert an e.g. matrix for an active-extrinsic rotation instead of an active-intrinsic one.
I compared the results for the rotated tensor and resulting strain with results from python (scipy, pytransform3D + matscipy), COMSOL, and OOF2 (where I used Bunge notation). The results match if I use an intrinsic and active rotation (matrix). 
Can you confirm that using active and intrinsic definition of the rotation for matrix-input is correct?

EDIT: I did compute passive-extrinsic Euler-angels out of my active and intrinsic rotation matrix and do get with these angles directly fed into MOOSE the same results.

**Reply by jessecarterMOOSE** [reply]
My understanding is that it is up to the user to determine what the proper orientation for their system is. If for instance, you have a precipitate phase embedded in a matrix, you could choose to either rotate the precipitate clockwise or rotate the matrix phase counterclockwise. Both are valid and you have the the ability to do either.
